### ISSUE TITLE
rewrite tonatona-db to only contain a connectionpool in the Shared data type

### ISSUE THREAD

**Issue by cdepillabout** [arow-oss/tonatona#11]
@arowM The following is from https://github.com/arow-oss/tonatona/pull/10#discussion_r201386340:

> Shared should only have pure values as its fields. So I recommend to change `TonaDb.Shared` to bellow.
> 
> ```haskell
> data Shared backend = Shared
>   { connectionPool :: ConnectionPool
>   }
> ```
> 
> Of course, if `ConnectionPool` is not actually needed to any DB connection, only Tonatona.Db.* modules should expose their own Shared data instead of exposing common Tonatona.Db.Shared.
> 
> What do you think about it?

I think this should work.

However, a `ConnectionPool` is actually a type alias:

```haskell
type ConnectionPool = Pool SqlBackend
```

Under the hood, this `Pool` record actually has members that are functions:

```haskell
data Pool a = Pool {
      create :: IO a
    , destroy :: a -> IO ()
    , numStripes :: Int
    , idleTime :: NominalDiffTime
    , maxResources :: Int
    , localPools :: V.Vector (LocalPool a)
    , fin :: IORef ()
    } deriving (Typeable)
```

It has functions for some of its fields.  Is something like this okay?

**Comment by cdepillabout** [comment]
Also, in the current sample code, it decides at runtime whether to use SQLite or PostgreSQL:

https://github.com/arow-oss/tonatona/blob/f4641d86d14e8c478d0d2cf3705d4c4d9478a925/tonatona-sample/src/Tonatona/Sample.hs#L139-L152

This is easy to do, since both `tonatona-db-postgresql` and `tonatona-db-sqlite` share the same `TonaDb.Shared` interface

I'm not sure how easy this would be to do if both `tonatona-db-postgresql` and `tonatona-db-sqlite` had different `Shared` types and different `TonaDbPostgresqlShared` or `TonaDbSqliteShared` type classes.

We might be able to get around this by declaring a common interface as described in https://github.com/arow-oss/tonatona/pull/10#discussion_r200815026:

```haskell
-- In `Tonatona.Email`
class TonaEmailInterface shared where
  send :: forall conf shared. shared -> Mail -> TonaM conf shared ()
-- (Does not declare `Shared` data type and `TonaEmailShared` type class in this module)

-- In `Tonatona.Email.Sendmail`
data Shared = Shared ()
init :: IO Shared
init = pure $ Shared ()
class TonaSendmailShared shared where
  shared :: shared -> Shared
instance (TonaSendmailShared shared) => TonaEmailInterface shared where
  send _shared = liftIO . renderSendMail

-- In `Tonatona.Email.Ses`
data Shared = Shared
    { token :: SesToken
    }
init :: Config -> IO Shared
init conf = getTokenForSes conf
class TonaSesShared shared where
  shared :: shared -> Shared
instance (TonaSesShared shared) => TonaEmailInterface shared where
  send shared' mail = undefined
    where
      Shared {token} = shared shared'
```

I'm not sure how well this would work, but it might be worth trying out!

**Comment by arowM** [comment]
> It has functions for some of its fields. Is something like this okay?

Good point!
I'm not sure whether we should limit fields of `Shared` to pure.

Now I have thoughts as bellow.

* When a function does not require preparation on startup (e.g., creating `connectionPool`), the function must not in `Shared` data type
    * Just simply declare it as a top level function as your refactored version of `Tonatona.Email.Sendmail.send`
* We should not expose `Shared` on root level modules (e.g., `Tonatona.Email`, `Tonatona.Db`, ...), but expose them on child modules (e.g., `Tonatona.Email.Sendmail`, `Tonatona.Email.Ses`, `Tonatona.Db.Postgres`, ...)
    * This is because there are no conviction that all of child modules has common shared data. For instance, non-persistent db libraries would not use `ConnectionPool`, so it is hard to decide root level `Shared` data.
* I don't know whether we should limit fields of `Shared` to pure or not

**Comment by arowM** [comment]
> I'm not sure how easy this would be to do if both tonatona-db-postgresql and tonatona-db-sqlite had different Shared types and different TonaDbPostgresqlShared or TonaDbSqliteShared type classes.

I guess it's not so hard by declaring `SharedDb` type as follows.

```
-- Shared

data Shared = Shared
  { tonaDb :: SharedDb
  , tonaLogger :: TonaLogger.Shared
  }

data SharedDb
  = SharedSqlite TonaDbSqlite.Shared
  | SharedPostgres TonaDbPostgres.Shared

instance Plug Config Shared where
  init conf = do
    let db =
          case dbToUse conf of
            Sqlite -> SharedSqlite <$> TonaDbSqlite.init conf stdoutLogger
            PostgreSQL -> SharedPostgres <$> TonaDbPostgres.init conf stdoutLogger
    Shared
      <$> db
      <*> TonaLogger.init stdoutLogger
```

**Comment by cdepillabout** [comment]
This has been fixed by #15.
